src/digraph.rs

```diff
@@ -2265,6 +2274,120 @@ impl PyDiGraph {
+                .iter()
+                .flat_map(|n| self.graph.edges(*n))
+                .filter_map(|edge| {
+                    let target_node = edge.target();
+                    if !nodes.contains(&target_node) {
+                        Some(target_node)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            // Now, if we can reach any of `nodes`, there exists a path from `nodes`
+            // back to `nodes` of length > 1, meaning contraction is disallowed.
+            let mut dfs = Dfs::from_parts(visit_next, self.graph.visit_map());
+            while let Some(node) = dfs.next(&self.graph) {
+                if nodes.contains(&node) {
+                    // we found a path back to `nodes`
+                    return false;
+                }
+            }
+
+            true
+        };
+
+        let mut indices_to_remove: HashSet<NodeIndex> =
+            to_replace.iter().map(|&n| NodeIndex::new(n)).collect();
+
+        if check_cycle.unwrap_or(self.check_cycle)
+            && !can_contract(&indices_to_remove)
+        {
+            return Err(DAGWouldCycle::new_err(
+                "Substitution would create cycle(s)",
+            ));
+        }
+
+        // Create new node.
+        let node_index = self.graph.add_node(obj);
+
+        // Sanitize new node index from user input.
+        indices_to_remove.remove(&node_index);
+
+        // Determine edges for new node.
+        let edges: Vec<(NodeIndex, NodeIndex, PyObject)> = {
+            let incoming_edges = indices_to_remove
+                .iter()
+                .flat_map(|&i| {
+                    self.graph.edges_directed(i, Direction::Incoming)
+                })
+                .filter_map(|edge| {
+                    let pred = edge.source();
+                    if !indices_to_remove.contains(&pred) {
+                        Some((pred, node_index, edge.weight().clone_ref(py)))
+                    } else {
+                        None
+                    }
+                });
+
+            let outgoing_edges = indices_to_remove
+                .iter()
+                .flat_map(|&i| {
+                    self.graph.edges_directed(i, Direction::Outgoing)
+                })
+                .filter_map(|edge| {
+                    let succ = edge.target();
+                    if !indices_to_remove.contains(&succ) {
+                        Some((node_index, succ, edge.weight().clone_ref(py)))
+                    } else {
+                        None
+                    }
+                });
+
+            incoming_edges.chain(outgoing_edges).collect()
+        };
+
+        for index in indices_to_remove {
+            self.graph.remove_node(index);
+        }
+
+        for (start, end, weight) in edges {
+            self._add_edge_no_checks(start, end, weight)?;
+        }
+
+        Ok(())
```

Comment:
```suggestion
        Ok(node_index)
```
since we should return the index of the new node.

Reply:
Done in 31be2d0.